src/Slacklinker/Extract/Types.hs

```diff
@@ -0,0 +1,50 @@
+module Slacklinker.Extract.Types where
+
+import Slacklinker.Import
+import Web.Slack.Conversation (ConversationId (..))
+import Web.Slack.Experimental.Blocks
+import Web.Slack.Experimental.Events.Types
+import Web.Slack.Files.Types (FileObject)
+import Web.Slack.Types qualified as Slack (UserId (..))
+
+data ExtractedMessageData = ExtractedMessageData
+  { channelType :: ChannelType
+  , channel :: ConversationId
+  , user :: Slack.UserId
+  , ts :: Text
+  , threadTs :: Maybe Text
+  , text :: Text
+  , blocks :: Maybe [SlackBlock]
+  , files :: Maybe [FileObject]
+  }
+  deriving stock (Show)
+
+class ExtractableMessage m where
+    extractData :: m -> ExtractedMessageData
+
+instance ExtractableMessage MessageEvent where
+    extractData MessageEvent {..} = 
+        ExtractedMessageData
+          { channelType
+          , channel
+          , user
+          , ts
+          , threadTs
+          , text
+          , blocks
+          , files
+          }
+
+instance ExtractableMessage BotMessageEvent where
+    extractData BotMessageEvent {..} = 
+        ExtractedMessageData
+          { channelType
+          , channel
+          -- a bit of a hack, but user ids and bot ids don't overlap
+          , user = Slack.UserId { unUserId = botId }
+          , ts
+          , threadTs
+          , text
+          , blocks
+          , files
+          }
```

Comment:
This is an extremely minor (and non-blocking) suggestion that I know @evanrelf will hate me for 😅, but:

```suggestion
    extractData MessageEvent {..} = ExtractedMessageData{..}

instance ExtractableMessage BotMessageEvent where
    extractData BotMessageEvent {..} = 
        ExtractedMessageData
          { -- a bit of a hack, but user ids and bot ids don't overlap
          , user = Slack.UserId { unUserId = botId }
          , ..
          }
```

Like, for me, using `RecordWildCards` for construction is actually the less evil use case for it (whereas using it for destruction is usually the more controversial thing).

Either way, I've always viewed slurping fields of the same name from one record to to another within a small scope to be one of the most compelling use cases for `RecordWildCards`.

Reply:
I still hate `RecordWildCards`, but this is among the least evil uses of it, so I approve.

Reply:
Done.